struct.rkt

```diff
@@ -62,3 +90,53 @@
 (provide *Range
          make-Range
          Range->hash)
+
+(struct FormattingOptions
+  (tabSize
+   insertSpaces
+   trimTrailingWhitespace
+   insertFinalNewline
+   trimFinalNewlines
+   key))
+
+(define/contract (make-FormattingOptions #:tabSize tabSize
+                                         #:insertSpaces insertSpaces
+                                         #:trimTrailingWhitespace [trimTrailingWhitespace undef-object]
+                                         #:insertFinalNewline [insertFinalNewline undef-object]
+                                         #:trimFinalNewlines [trimFinalNewlines undef-object]
+                                         #:key [key undef-object])
+  (-> #:tabSize uinteger?
+      #:insertSpaces boolean?
+      #:trimTrailingWhitespace (undef/c boolean?)
+      #:insertFinalNewline (undef/c boolean?)
+      #:trimFinalNewlines (undef/c boolean?)
+      #:key (undef/c hash?)
+      FormattingOptions?)
+
+  (FormattingOptions tabSize
+                     insertSpaces
+                     trimTrailingWhitespace
+                     insertFinalNewline
+                     trimFinalNewlines
+                     key))
+
+(define (jsexpr->FormattingOptions js)
+  (with-handlers ([exn:fail? (λ (_) request-err-object)])
+    (make-FormattingOptions #:tabSize (hash-ref js 'tabSize)
+                            #:insertSpaces (hash-ref js 'insertSpaces)
+                            #:trimTrailingWhitespace (hash-ref js 'trimTrailingWhitespace undef-object)
+                            #:insertFinalNewline (hash-ref js 'insertFinalNewline undef-object)
+                            #:trimFinalNewlines (hash-ref js 'trimFinalNewlines undef-object)
+                            #:key (hash-ref js 'key undef-object))))
+
+(define-match-expander as-FormattingOptions
+  (lambda (stx)
+    (syntax-parse stx
+      [(_ name)
+       #'(and (? hash?)
+              (app jsexpr->FormattingOptions (and name (not (? request-err?)))))])))
```

Comment:
Add an example in comment for this pattern macro would be great

Reply:
ok